File: hashing2/src/subarraysumequalsk.java
import java.util.HashMap;

class subarraysumequalsk {
    public int subarraySum(int[] nums, int k) {
        //initializing hashmap,count,cuurentsum.
        HashMap<Integer,Integer> hashmap = new HashMap<>();
        int count=0;
        int currentsum=0;
        hashmap.put(0,1);


        for(int x: nums){
            currentsum+=x;
            int compliment=currentsum-k;
            //checking compliment in hashmap
            if(hashmap.containsKey(compliment)){
                count+=hashmap.get(compliment);
            }
            //checking currentsum in hashmap
            if(hashmap.containsKey(currentsum)){
                hashmap.put(currentsum,hashmap.get(currentsum)+1);
            }
            //adding current sum to hashmap
            else{
                hashmap.put(currentsum,1);
            }

        }
        return count;
    }
}

//time complexity=o(n)
//space complexity=o(n)

File: SubArraySumEqualsK.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Algorithms
{
    // Did this code successfully run on Leetcode : No, but when I code the same in Java, it runs. Looks like
    // there is an issue with the Leetcode compiler with C# code base
    // Any problem you faced while coding this : No
    public class SubArraySumEqualsK
    {
        //Time Complexity - O(n) - As we are iterating through an array
        // Space Complexity - O(n) - As we are using Extra space by using Dictionary
        public int SubarraySum(int[] nums, int k)
        {
            int rSum = 0; int count = 0;
            var dict = new Dictionary<int, int>();
            dict.Add(0, 1);
            for (int i = 0; i < nums.Length; i++)
            {
                rSum = rSum + nums[i];
                // If the difference between running sum and the total(k) is found in the dictionary, that 
                // means that there will be a subArray that sums up to k before this index
                if (dict.ContainsKey(rSum - k))
                {
                    count = count + dict.GetValueOrDefault(rSum - k, 0);
                }
                // Add runningSum as the index and the value (number of times runningSum is found during iteration) 
                dict.Add(rSum, dict.GetValueOrDefault(rSum, 0) + 1);
            }
            return count;
        }
    }
}
